Add configurable file-existence and HTTP health checks

aaronlehmann opened the pull request:
Add a section to the config file called "health". Within this section,
"filecheckers" and "httpcheckers" list checks to run. Each check
specifies a file or URI, a time interval for the check, and a threshold
specifying how many times the check must fail to reach an unhealthy
state.

Document the new options in docs/configuration.md.

Add unit testing for both types of checkers. Add an UnregisterAll
function in the health package to support the unit tests, and an
Unregister function for consistency with Register.

Fix a string conversion problem in the health package's HTTP checker.

Fixes #271.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "configurable-health-checks" git@github.com:aaronlehmann/distribution.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

RichardScothern commented on registry/handlers/app.go:
```diff
@@ -212,10 +216,38 @@ func NewApp(ctx context.Context, configuration configuration.Configuration) *App
 // if called twice in the same process.
 func (app *App) RegisterHealthChecks() {
-	health.RegisterPeriodicThresholdFunc("storagedriver_"+app.Config.Storage.Type(), 10*time.Second, 3, func() error {
+	health.RegisterPeriodicThresholdFunc("storagedriver_"+app.Config.Storage.Type(), defaultCheckInterval, 3, func() error {
```
Is it worth making the storage checker configurable as per file and http checkers?

aaronlehmann replied:
I avoided touching this to keep the scope of the PR limited, but it wouldn't be much work to add some options here. Does something like this look good?

    health:
      storagedriver:
        enabled: true
        interval: 10s
        threshold: 3

Without a "storagedriver" section, this would default to being enabled.

stevvooe replied:
@aaronlehmann This would probably be the most useful.

RichardScothern replied:
That looks fine to me.  Rather than enable it by default, could you just add it to the example configurations?

stevvooe commented on registry/handlers/health_test.go:
```diff
@@ -0,0 +1,200 @@
+
+	"github.com/docker/distribution/configuration"
+	"github.com/docker/distribution/health"
+	"golang.org/x/net/context"
```
Please use `distribution/context`.

aaronlehmann replied:
Several existing tests use `golang.org/x/net/context`. Should I update these while I'm at it?

stevvooe replied:
@aaronlehmann If it is unobtrusive. For new stuff, use `distribution/context`.

stevvooe commented on health/health.go:
```diff
@@ -170,6 +170,20 @@ func Register(name string, check Checker) {
 }
 
+// Unregister removes the named checker.
+func Unregister(name string) {
```
I'm not a fan of `Unregister`. This pattern doesn't exist in the stdlib. Can we test the health checkers without registration?

aaronlehmann replied:
I suppose it would be possible to skip the registration and just call the checker directly, but if we did that, the unit test would be testing something very different from what the registry actually does. It wouldn't check interpretation of the config structure or registration of the checks, only the fact that the health package's HTTP check or file check works as promised.

If we do register the checks, we need some way to remove registrations, so subsequent tests don't interfere with each other. Without `UnregisterAll`, there's a panic when the second test tries to call `RegisterHealthChecks`.

It would be possible to combine these two tests into one, but that feels less flexible and less robust. If we ever end up in a situation where another test registers a health check, we'd have the same problem.

I'm open to suggestions on how to deal with this.

stevvooe replied:
@aaronlehmann I had this same problem when adding the driver health check. Generally, we are following the model of global health check registry to maintain uniqueness for the process. Obviously, this is at odds with unit testing. We may want to look at the `flag` package or the `net/http` package for testing feedback on global registration patterns.

What if we change the signature of `RegisterHealthChecks` to `RegisterHealthChecks(registerFn func(name string, check Checker))`? The main file passes in `health.Register` and the unit tests pass in their own registry than can be checked for correctness.

The other option would be to export a `health.Registry` object that can be used to manage registrations. `health.Register` calls into the `DefaultRegistry` but unit tests can define a separate one. This is more work but probably would provide a better result.

stevvooe commented on registry/handlers/health_test.go:
```diff
@@ -0,0 +1,200 @@
+	close(stopFailing)
+
+	<-time.After(2 * interval)
+	resp, err := http.Get(debugServer.URL + "/debug/health")
```
We shouldn't be testing the http interface of the health package.

aaronlehmann replied:
You have a good point; calling `health.CheckStatus` directly would be more straightforward.

That said, since these tests already exist in this form and provide more end-to-end coverage by accessing the HTTP interface, how would you feel about leaving them this way to provide this extra coverage?

stevvooe replied:
My only issue is that these tests are harder to maintain over time and I suspect we will make changes to the http interface.

It is probably okay to leave this since it has already been done.

stevvooe commented on docs/configuration.md:
```diff
@@ -195,6 +195,15 @@ information about each option that appears later in this page.
+        - file: /path/to/checked/file
+          interval: 10s
+          threshold: 3
+      httpcheckers:
```
`http` instead of `httpcheckers. `file` instead of `filecheckers`.

stevvooe commented:
What is the use case of an http health checker? I could imagine something that puts the registry into 503 if a notification endpoint is down. Should we integrate that closer with the notifications declaration, automatically registering a health check or does it make sense to have to configure it separately? I think we can make a solid argument for the latter.

RichardScothern commented:
An HTTP health checker allows a registry to monitor the health of its notification endpoints, which makes sense to me.  However the reaction to this (disabling itself) doesn't really make too much sense, especially given that all registry instances may notify the same endpoints.

It would be nice if the healthchecker could allow an action to be configured on a failing health check, but that's way out of scope for this PR.

aaronlehmann commented:
I don't have a good understanding of the use case for the HTTP health checker. Using it to check a notification endpoint would make sense. I do think this should be configured separately from the notification itself, though.

noxiouz commented on health/health.go:
```diff
@@ -170,6 +170,20 @@ func Register(name string, check Checker) {
+func Unregister(name string) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	delete(registeredChecks, name)
```
Is Unregister expected to call in runtime? I'm not sure if it's necessary, but Checks could leave various footprints, (for example, goroutines). May be we should add Close method to Checker interface?

aaronlehmann replied:
I'm not expecting anything to call it at runtime. Currently it's only used in tests. But it's a good point that resources would leak with the current implementation.

aaronlehmann commented:
Renamed yaml fields; added `storagedriver` section to config.

stevvooe commented:
> However the reaction to this (disabling itself) doesn't really make too much sense, especially given that all registry instances may notify the same endpoints.

Should we allow this to be configurable? For example, only registry `alive` if it is configured?

aaronlehmann commented:
Added a concept of `Registry` objects to `health` package and modified the unit tests to use custom registries instead of unregistering checks.

stevvooe commented on docs/configuration.md:
```diff
@@ -195,6 +195,19 @@ information about each option that appears later in this page.
+          interval: 10s
+          threshold: 3
+      http:
+        - uri: http://server.to.check/must/return/200
```
HTTP health should allow a configurable response code.

aaronlehmann replied:
> HTTP health should allow a configurable response code.

Good idea. I'm also adding a timeout parameter.

stevvooe replied:
Notifications has a good subset of http options:

```
// Endpoint describes the configuration of an http webhook notification
// endpoint.
type Endpoint struct {
	Name      string        `yaml:"name"`      // identifies the endpoint in the registry instance.
	Disabled  bool          `yaml:"disabled"`  // disables the endpoint
	URL       string        `yaml:"url"`       // post url for the endpoint.
	Headers   http.Header   `yaml:"headers"`   // static headers that should be added to all requests
	Timeout   time.Duration `yaml:"timeout"`   // HTTP timeout
	Threshold int           `yaml:"threshold"` // circuit breaker threshold before backing off on failure
	Backoff   time.Duration `yaml:"backoff"`   // backoff duration
}
```

aaronlehmann replied:
I could add `headers` as well. Do you think there are any others worth adding for the health check?

stevvooe commented on docs/configuration.md:
```diff
@@ -195,6 +195,19 @@ information about each option that appears later in this page.
+      file:
+        - file: /path/to/checked/file
+          interval: 10s
+          threshold: 3
```
Threshold doesn't really make sense for a file check.

aaronlehmann commented:
Added a TCP checker (see discussion in https://github.com/docker/distribution/pull/678). Also added `timeout`, `statuscode`, and `headers` parameters to the HTTP checker, and removed `threshold` from the file checker. I believe this is ready to go unless there's more feedback.

stevvooe commented on health/health.go:
```diff
@@ -159,33 +179,39 @@ func CheckStatus() map[string]string { // TODO(stevvooe) this needs a proper typ
-	_, ok := registeredChecks[name]
+// overwrites to a specific check status. registry may be nil to use the
+// default registry.
+func Register(registry *Registry, name string, check Checker) {
```
Don't change the signature of this function: I really don't like nil parameters hanging about. This should just be a method on `Registry`.

aaronlehmann replied:
I modeled this after `http.ListenAndServe`. It accepts a nil parameter for the handler to use the default mux. If this was a method on `Registry`, it would need to be used like

    health.DefaultRegistry.Register(...)

...which seems awkward to me - I would rather users never have to touch DefaultRegistry directly.

stevvooe replied:
Users should still call `health.Register`, which is the same as `health.DefaultRegistry.Register` without the first argument.

stevvooe commented on registry/handlers/app.go:
```diff
@@ -211,11 +215,72 @@ func NewApp(ctx context.Context, configuration configuration.Configuration) *App
-		_, err := app.driver.List(app, "/") // "/" should always exist
-		return err                          // any error will be treated as failure
-	})
+func (app *App) RegisterHealthChecks(healthRegistry *health.Registry) {
```
`RegisterHealthChecks(healthRegistries ...*health.Registry)` allows you to omit registry in default case.

aaronlehmann replied:
That's an interesting idea, but to me it implies that the function can register the checks in multiple health registries. While we could make it support that, I can't see a use case where it would make sense. So I'm not sure this is a good fit for the situation.

stevvooe replied:
More importantly, it allows you to all it like this `RegisterHealthChecks()`. Did I mention I really hate `nil` function parameters?

aaronlehmann commented:
Changed so that `Register`/`CheckStatus`/`RegisterPeriodicFunc`/etc. are available as `Registry` methods as well as standalone functions, and did the varargs thing for `RegisterHealthChecks`.

stevvooe commented:
LGTM

RichardScothern commented:
LGTM
